### DISCUSSION TITLE
[QUESTION]Splitting large document and bucketing

### DISCUSSION THREAD

**Discussion by shafiqabedin** [NVIDIA/Megatron-LM#1090] [Q&A]
I asked this question in the discussion section but did not receive any response so asking here with a little bit more details. I am trying to figure out if bucketization is done (or can be done) for model training. By "bucketization" I am referring to batching based on similar sequence length [(https://torchtext.readthedocs.io/en/latest/data.html#bucketiterator).](https://torchtext.readthedocs.io/en/latest/data.html#bucketiterator). The motivation is that I have documents with very large text and I want pick a splitting schema (which may create a lot of samples with small number of tokens and bucketize them). That brings me to the second question which is - is splitting supported in Megatron?

Any answer would be much appreciated - Thanks.

**Comment by Advait251206** [comment]
Yes — this is something Megatron-LM can handle, but **not in the same “out-of-the-box BucketIterator + automatic splitting” way that torchtext does**. You need to understand how Megatron’s data pipeline is designed.

---

## 1. Does Megatron-LM support bucketization?

### ✔ Yes — but indirectly

Megatron-LM achieves “bucket-like behavior” through:

- **dynamic sequence packing**
- **length-based batching (optional depending on dataset builder)**
- **token stream construction instead of fixed samples**

However:

> It does NOT use a PyTorch-style BucketIterator abstraction.

Instead, it builds a **contiguous token dataset** and then chunks it.

---

## 2. How Megatron actually batches data

Megatron training data pipeline is:

```
Documents → Tokenization → Concatenation → Fixed-length chunking → Batches
```

So batching is based on:

- fixed `seq_length`
- continuous token stream slicing

NOT:

- grouping similar-length documents at runtime

---

## 3. Is explicit bucketization supported?

### ❌ Not natively like torchtext

Megatron does not:

- group samples by length dynamically
- maintain multiple buckets per batch iterator
- reorder dataset online based on sequence length

---

## 4. What Megatron uses instead

### ✔ Token packing (closest equivalent)

If enabled in dataset preprocessing:

- short documents are concatenated
- long documents are split
- sequences are filled up to `seq_length`

This achieves:

> high GPU utilization without explicit bucketing

---

## 5. Your use case: large documents + splitting

You said:

> “split large documents into many small samples”

This is already how Megatron expects data to be structured.

### ✔ You should:

Preprocess offline:

- split long docs into chunks (e.g., 1K–4K tokens)
- optionally overlap chunks (sliding window)
- store as flat token stream or document list

Then Megatron will:

- pack them into fixed-length sequences automatically

---

## 6. Can Megatron do splitting automatically?

### ❌ Not intelligently

It does NOT:

- detect semantic boundaries
- split documents by meaning
- optimize chunking per-document

You must do this **before training**.

---

## 7. Why true bucketization is not used

Bucketization (like torchtext) improves:

- padding efficiency
- per-batch uniformity

But Megatron prefers:

### ✔ global token streaming

Because it enables:

- deterministic training
- efficient tensor parallel execution
- simple distributed sharding
- stable memory layout

---

## 8. What about “length-aware batching”?

Some Megatron variants can:

- reduce padding waste via packing
- optionally sort by length (offline preprocessing step)

But this is:

> preprocessing-time sorting, not runtime bucketing

---

## 9. Recommended approach (best practice)

For your case:

### ✔ Step 1: split documents offline

Example:

- max chunk size = 2048 or 4096 tokens
- overlap = 128–256 tokens (optional)

---

### ✔ Step 2: shuffle chunks globally

- avoid ordering bias
- ensures randomness

---

### ✔ Step 3: feed into Megatron as token stream

Let Megatron handle:

- batching
- packing
- distributed partitioning

---

## 10. Key distinction

| Feature | torchtext BucketIterator | Megatron-LM |
|--------|-------------------------|--------------|
| Dynamic length batching | ✔ | ❌ |
| Runtime bucketing | ✔ | ❌ |
| Offline packing | optional | ✔ core design |
| Token stream training | ❌ | ✔ primary model |

---

## 11. Bottom line

- ❌ Megatron does not support runtime bucketization like torchtext
- ✔ It uses token-stream packing instead
- ✔ Splitting large documents must be done in preprocessing
- ✔ Efficiency is achieved via fixed-length chunking, not dynamic batching

---

If this answer helped or pointed you in the right direction, I'd appreciate it if you could mark it as the accepted answer so it's easier for others with the same issue to find.

Also, if you found my contribution useful, I'd appreciate it if you could check out my GitHub profile, follow me, and star any repositories you find interesting.

GitHub: https://github.com/Advait251206
